### ISSUE TITLE
[Wave4] embed_indices.py 実装 (gte-Qwen2 per-CIK shard embedding, act-105)

### ISSUE THREAD

**Issue by YH-05** [YH-05/quants#3947]
## 概要

`05_embedding.ipynb` (9 銘柄スコープの実装) の Cell 4-6 ロジックをモジュール化した embedding 生成 CLI を実装する。NAS の `chunks/indices_v1/chunks_cik{cik:010d}.parquet` を per-CIK で順次読み込み、gte-Qwen2-1.5B-instruct (MPS bfloat16) で encode し、`embeddings/indices_v1/embeddings_cik{cik:010d}.npy` (float32 1536 dim, L2-normalized) + `chunks_meta_cik{cik:010d}.parquet` を per-CIK 分離で出力する。NaN マーカーによる resume、CIK 単位 Checkpoint、失敗バッチは `unresolved_chunks.jsonl` (HF1 default: リトライなし)。

議論メモ該当箇所: `docs/plan/2026-05-25_discussion-indices-v1-pipeline.md` セクション「2. 展開計画」(行 35-42) および「7. 規模試算」(行 78-86)

## 関連 ActionItem / Decision

- ActionItem: **act-2026-05-25-105**
- Decision:
  - **dec-2026-05-25-102**: スコープ = chunks 生成 + gte-Qwen2-1.5B embedding 生成まで
  - **dec-2026-05-25-105**: CLI スクリプト中心実装
- HF1 確定: embedding 保存形式 = per-CIK で .npy + meta parquet 分離 / リトライなし (失敗バッチは jsonl 記録) / rate limiter 不要

## 実装内容

`notebook/FILING_NLP/pipeline/embed_indices.py` (新規, 10-12KB)

### CLI

```bash
python -m notebook.FILING_NLP.pipeline.embed_indices \
  --run-id indices_v1 \
  --membership /Volumes/personal_folder/Quants/FILING_NLP_v2/index_membership/membership_indices_v1.parquet \
  --index-filter in_spx \
  [--batch-size 16] [--max-length 512] [--device mps] [--dtype bfloat16] \
  [--checkpoint-every-n-batches 10] [--force-reencode]
```

### 関数構成

```python
def _load_model(model_id: str, device: str, dtype: str) -> tuple[AutoModel, AutoTokenizer]:
    """AutoModel.from_pretrained(model_id, trust_remote_code=True, dtype=torch.bfloat16,
    device_map='mps', low_cpu_mem_usage=True) + use_cache=False で DynamicCache 互換性回避"""

def last_token_pool(last_hidden_states, attention_mask):
    """right/left padding 両対応 (05_embedding.ipynb 流用)"""

def encode_texts(
    texts: list[str], model, tokenizer, batch_size: int, max_length: int
) -> np.ndarray:
    """バッチ encode → last_token_pool → F.normalize(p=2, dim=1) → float32 numpy (N, 1536)"""

def embed_cik(
    cik: int,
    chunks_parquet: Path,
    out_npy: Path,
    out_meta: Path,
    model,
    tokenizer,
    batch_size: int,
    force_reencode: bool,
    checkpoint_every_n_batches: int,
) -> dict:
    """1 CIK の chunks を読み込み embedding 生成。
    - 既存 .npy があれば NaN マーカー (np.isnan(arr).any(axis=1)) で未処理 index 取得
    - force_reencode=True で全件再生成
    - checkpoint_every_n_batches バッチごとに np.save で永続化
    - 失敗バッチは unresolved_chunks.jsonl に append-only
    Returns: {'n_chunks': int, 'n_embedded': int, 'elapsed_sec': float}"""

def main() -> None:
    """argparse → membership ロード + index_filter → model ロード →
    CIK 単位ループ (Checkpoint resume) → embed_cik 呼び出し → summary.json 保存"""

if __name__ == "__main__":
    main()
```

### 保存形式

- `embeddings/indices_v1/embeddings_cik{cik:010d}.npy`: shape (N_chunks, 1536), float32, L2-normalized
- `embeddings/indices_v1/chunks_meta_cik{cik:010d}.parquet`: chunk_id (zfill 6) + その他メタ
- `checkpoints/indices_v1_embed_progress.json`: CIK 単位 completed 記録 (`INDICES_V1_EMBED_PROGRESS_PATH`)
- `logs/indices_v1_embed_errors.jsonl`: 失敗バッチ (cik / batch_idx / error 記録)
- `logs/unresolved_chunks.jsonl`: バッチ単位失敗の append-only 記録

### HF1 確定方針

- **リトライなし**: バッチ単位で失敗したら `unresolved_chunks.jsonl` に記録して継続 (HF1 default 案A)
- **rate limiter なし**: SEC 通信なし、HF cache 前提 (HF1 default 案A)
- **NaN マーカー resume**: `np.full((N, 1536), np.nan, dtype=np.float32)` 初期化、`np.isnan(arr).any(axis=1)` で再開位置取得

### pilot100 chunks での先行検証

SPX chunks 完走 (task-4) 前でも、`pilot100` の `chunks_cik*.parquet` で `embed_cik()` の動作検証可能 (Wave 4 を Wave 2 と並列化するオプション)。

## 受け入れ条件

- [ ] `notebook/FILING_NLP/pipeline/embed_indices.py` が作成され、5 関数 (`_load_model`, `last_token_pool`, `encode_texts`, `embed_cik`, `main`) を持つ
- [ ] CLI `python -m notebook.FILING_NLP.pipeline.embed_indices --help` が動作
- [ ] pilot100 chunks (またはタスク 4 で生成された SPX chunks の 1-2 CIK) で `embed_cik()` を実行し、`embeddings_cik{cik:010d}.npy` shape が `(N, 1536)` で生成される
- [ ] 生成された .npy の L2 ノルムが ~1.0 (`np.linalg.norm(arr, axis=1)` で確認)
- [ ] `--force-reencode` なしで再実行した場合、NaN マーカー部分のみ再エンコード
- [ ] `--force-reencode` で全件再生成
- [ ] `checkpoints/indices_v1_embed_progress.json` で CIK 単位 resume が機能 (中断 → 再起動で完了 CIK は skip)
- [ ] バッチ失敗時に `unresolved_chunks.jsonl` に append-only で記録 (MPS OOM 等のシミュレーションで確認)
- [ ] `make check-all` 成功 (format / lint / typecheck)

## 依存タスク

- 依存: **task-1** (config.py の EMBEDDING_* 定数), **task-4** (SPX chunks 本番データ) — ただし pilot100 chunks での先行検証なら task-4 待ち不要
- ブロックする: task-6 (SPX embedding 起動)

## 参考リソース

- 議論メモ: `docs/plan/2026-05-25_discussion-indices-v1-pipeline.md` 行 35-42, 78-86, 106
- ベース実装: `notebook/FILING_NLP/05_embedding.ipynb` (Cell 4-6 が encode_texts + resume 本体)
- 再利用クラス: `notebook/FILING_NLP/pipeline/runner.py` の `Checkpoint` クラス (checkpoint_path を `INDICES_V1_EMBED_PROGRESS_PATH` に向けるだけで流用可)
- HF cache パターン: `notebook/FILING_NLP/_helpers.py` の `setup_hf_cache()` + `get_device()`
- HF1 回答: `user-answers.json` (embedding_storage / error_handling / rate_limit_for_embedding)

## リスクと緩和策

- **リスク**: gte-Qwen2-1.5B-instruct の MPS bfloat16 サポートが PyTorch バージョン依存。9 銘柄では検証済だが SPX 503 銘柄 ~10h 連続動作の安定性は未検証
  - **緩和**: CIK 単位 Checkpoint + バッチ単位永続化 (`--checkpoint-every-n-batches 10`) で異常終了後も resume 可能。MPS OOM 頻発時は `--batch-size 8` への引き下げで CLI で即時対応
- **リスク**: 既存 src/embedding (Gemini/ChromaDB 用) との HF cache 衝突
  - **緩和**: HF_HOME / HUGGINGFACE_HUB_CACHE 環境変数を `_helpers.py:setup_hf_cache()` 流儀で明示設定し、衝突回避

## 関連

- 計画書: docs/project/project-107/project.md
- GitHub Project: [#114](https://github.com/users/YH-05/projects/114)
- 元議論: docs/project/project-107/original-plan.md
- Neo4j Discussion: disc-2026-05-25-indices-v1-pipeline
